File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode], lower=float("-inf"), upper=float("inf")) -> bool:
        if not root:
            return True
        if not (lower < root.val < upper):
            return False

        if not root.left and not root.right:
            return True
        if not root.left and root.right:
            return root.val < root.right.val and self.isValidBST(root.right, lower=root.val, upper=upper)
        if root.left and not root.right:
            return root.left.val < root.val and self.isValidBST(root.left, lower=lower, upper=root.val)
        
        return root.left.val < root.val < root.right.val and self.isValidBST(root.left, lower=lower, upper=root.val) and self.isValidBST(root.right, lower=root.val, upper=upper)

File: Data Structures & Algorithms/generate-parentheses/submission-4.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        std::vector<std::vector<std::string>> cache(n + 1);
        cache[0].push_back("");

        for (int i = 1; i < n + 1; ++i) {
            for (int l = 0; l < i; ++l) {
                const int r = i - 1 - l;

                for (auto left : cache[l]) {
                    for (auto right : cache[r]) {
                        cache[i].push_back(
                            '(' + left + ')' + right
                        );
                    }
                }
            }
        }

        return cache[n];
    }
};
